modules/core/test/test_hasnonzero.cpp

```diff
@@ -0,0 +1,298 @@
+        case CV_16U: { if (!src.at<ushort>(i, j)) {src.at<ushort>(i, j) = cv::randu<ushort>(); n += (src.at<ushort>(i, j) > 0);} break; }
+        case CV_16S: { if (!src.at<short>(i, j)) {src.at<short>(i, j) = cv::randu<short>(); n += abs(sign(src.at<short>(i, j)));} break; }
+        case CV_32S: { if (!src.at<int>(i, j)) {src.at<int>(i, j) = cv::randu<int>(); n += abs(sign(src.at<int>(i, j)));} break; }
+        case CV_32F: { if (fabs(src.at<float>(i, j)) <= eps_32) {src.at<float>(i, j) = cv::randu<float>(); n += (fabs(src.at<float>(i, j)) > eps_32);} break; }
+        case CV_64F: { if (fabs(src.at<double>(i, j)) <= eps_64) {src.at<double>(i, j) = cv::randu<double>(); n += (fabs(src.at<double>(i, j)) > eps_64);} break; }
+
+        default: break;
+        }
+    }
+
+}
+
+void CV_HasNonZeroTest::generate_src_stat_data(cv::Size size, int type, int distribution)
+{
+    src.create(size, CV_MAKETYPE(type, 1));
+
+    double mean = 0.0, sigma = 1.0;
+    double left = -1.0, right = 1.0;
+
+    RNG& rng = ts->get_rng();
+
+    if (distribution == RNG::NORMAL)
+        rng.fill(src, RNG::NORMAL, Scalar::all(mean), Scalar::all(sigma));
+    else if (distribution == RNG::UNIFORM)
+        rng.fill(src, RNG::UNIFORM, Scalar::all(left), Scalar::all(right));
+}
+
+bool CV_HasNonZeroTest::get_has_non_zero()
+{
+    bool result = false;
+
+    for (int i = 0; i < src.rows; ++i)
+        for (int j = 0; j < src.cols; ++j)
+        {
+            if (current_type == CV_8U) result |= (src.at<uchar>(i, j) > 0);
+            else if (current_type == CV_8S) result |= (abs(sign(src.at<char>(i, j)))>0);
+            else if (current_type == CV_16U) result |= (src.at<ushort>(i, j) > 0);
+            else if (current_type == CV_16S) result |= (abs(sign(src.at<short>(i, j)))>0);
+            else if (current_type == CV_32S) result |= (abs(sign(src.at<int>(i, j))) > 0);
+            else if (current_type == CV_32F) result |= (fabs(src.at<float>(i, j)) > eps_32);
+            else result |= (fabs(src.at<double>(i, j)) > eps_64);
+        }
+
+    return result;
+}
+
+void CV_HasNonZeroTest::print_information(bool right, bool result)
+{
+    cout << endl; cout << "Checking for the work of hasNonZero function..." << endl; cout << endl;
+    cout << "Type of Mat: ";
+    switch (current_type)
+    {
+    case 0: {cout << "CV_8U"; break;}
+    case 1: {cout << "CV_8S"; break;}
+    case 2: {cout << "CV_16U"; break;}
+    case 3: {cout << "CV_16S"; break;}
+    case 4: {cout << "CV_32S"; break;}
+    case 5: {cout << "CV_32F"; break;}
+    case 6: {cout << "CV_64F"; break;}
+    default: break;
+    }
+    cout << endl;
+    cout << "Number of rows: " << src.rows << "   Number of cols: " << src.cols << endl;
+    cout << "True has non zero elements: " << right << "   Result: " << result << endl;
+    cout << endl;
+}
+
+void CV_HasNonZeroTest::run(int)
+{
+    const size_t N = 1500;
+
+    for (int k = 1; k <= 3; ++k)
```

Comment:
Why do you need loop, if all k are handled in swtch?

Reply:
- I propose to reduce switch and extract common part. No need to repeat the same several times.
- No need to handle results and print status manually. Use it parameter and call the function with particular k in dedicated test instance. Result will be obvious from GTest output.

Reply:
I just took the code from test_countnonzero, this is the exact same scheme.
What is the best option : make that scheme evolve for cv_hasNonZero(), or keep homogeneity between those two unit tests ?

Reply:
test_countnonzero was written in pre-GTest epoch. No need to do the same. Feel free to use Gtest in more efficient way.

Reply:
Ok, I will refactor the whole hasNonZero tests to modern code (within a few days )

Reply:
Done (and it exposed a bug that I could fix)